shell/platform/darwin/ios/framework/Source/FlutterViewController.mm

```diff
@@ -310,8 +318,37 @@ - (void)pushRoute:(NSString*)route {
 
 #pragma mark - Loading the view
 
+static UIView* GetViewOrPlaceholder(UIView* existing_view) {
+  if (existing_view) {
+    return existing_view;
+  }
+
+  auto placeholder = [[[UIView alloc] init] autorelease];
+
+  placeholder.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
+  placeholder.backgroundColor = UIColor.whiteColor;
+  placeholder.autoresizesSubviews = YES;
+
+  // Only add the label when we know we have failed to enable tracing (and it was necessary).
+  // Otherwise, a spurious warning will be shown in cases where an engine cannot be initialized for
+  // other reasons.
+  if (flutter::GetTracingResult() == flutter::TracingResult::kDisabled) {
+    auto messageLabel = [[[UILabel alloc] init] autorelease];
+    messageLabel.numberOfLines = 0u;
+    messageLabel.textAlignment = NSTextAlignmentCenter;
+    messageLabel.autoresizingMask =
+        UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
+    messageLabel.text =
+        @"Flutter application in debug mode can only be launched from Flutter tooling "
+        @"or Xcode. Use profile or release modes for launches from the home-screen.";
+    [placeholder addSubview:messageLabel];
+  }
+
+  return placeholder;
+}
+
 - (void)loadView {
-  self.view = _flutterView.get();
+  self.view = GetViewOrPlaceholder(_flutterView.get());
```

Comment:
This seems a bit derived. i.e. it's not a very clear, direct causation chain between the EnableTracingIfNecessary failing, the engine being nil, the view being nil etc to the ultimate text shown: "Flutter application in debug mode can only be launched from Flutter tooling or Xcode". If we show a specific solution, let's double check a specific diagnosis, i.e. engine or view is nil AND GetTracingResult failed.

Reply:
Maybe checking the engine directly is a bit cleaner. Otherwise there's still an implementation leak here where the vc knows that the view implementation returns nil when the delegate is nil.

Reply:
Never mind, you did inside the placeholder method. Are we adding new vectors to end up in a white placeholder view which I think shouldn't be possible before?

Reply:
> let's double check a specific diagnosis,

This check is being done in the method [`GetViewOrPlaceholder`](https://github.com/flutter/engine/pull/20980/files#diff-366c592b71736b6c5519e04346bf325cR335) method. The comment above that check explicitly mentions the case you described.

Reply:
Are we adding new vectors to end up in a white placeholder view which I think shouldn't be possible before?

I couldn't think of any but wanted to only add the message if we were sure of the cause of the failure.